File: BaseLibrary/protocol/ProtocolDouble.cs
using System;

namespace BaseLibrary.protocol {
    /// <summary>
    /// Represent double witch is safe for send  by network. Its number is save in int type with 3 decimal places (that is mean that 1000 [in ProtocolDouble] is equals to 1 and 12 [in ProtocolDouble] is equals to 0.012)
    /// </summary>
    public class ProtocolDouble {
        public static readonly ProtocolDouble ZERO = new ProtocolDouble(0.0);
        public static readonly ProtocolDouble _0 = ZERO;
        public static readonly ProtocolDouble _100 = new ProtocolDouble(100.0);

        /// <summary>
        /// Quotient for multiply double.
        /// </summary>
        public const double MULTIPLE = 1000.0; // it has to be power of ten

        public static bool TryParse(string s, out ProtocolDouble result) {
            bool success = int.TryParse(s, out int v);
            result = new ProtocolDouble(v / MULTIPLE);
            return success;
        }

        public static explicit operator ProtocolDouble(double d) {
            return new ProtocolDouble(d);
        }

        public static implicit operator double(ProtocolDouble pd) {
            return pd.value / MULTIPLE;
        }

        public int value { get; private set; }

        public ProtocolDouble(double d) {
            value = (int)(d * MULTIPLE);
        }

        public ProtocolDouble(ProtocolDouble d) {
            value = d.value;
        }

        public string Serialize() {
            return value.ToString();
        }

        public override string ToString() {
            return (value).ToString();
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/protocol/ProtocolDouble.java
package BaseLibrary.communication.protocol;

public class ProtocolDouble {
	public static final ProtocolDouble ZERO = new ProtocolDouble(0.0);
	public static final ProtocolDouble _0 = ZERO;
	public static final ProtocolDouble _100 = new ProtocolDouble(100.0);

	public static final double MULTIPLE = 1000.0; // it have to be power of ten

	/**
	 * 
	 * @param s
	 * @return
	 * @throws NullPointerException if string is null
	 * @throws NumberFormatException if the string does not contain a parsable ProtocolDouble.
	 */
	public static ProtocolDouble Parse(String s) {
		int v = Integer.parseInt(s);
		return new ProtocolDouble(v);
	}
	
	
	/**
	 * @param s
	 * @return null if fail
	 */
	public static ProtocolDouble TryParse(String s) {
		ProtocolDouble result = null;
		try {
			int v = Integer.parseInt(s);
			result = new ProtocolDouble(v);
		} catch (NumberFormatException e) {
			result = null;
		}
		return result;
	}

	private int value;

	public ProtocolDouble(double d) {
		value = (int) (d * MULTIPLE);
	}

	public ProtocolDouble(ProtocolDouble d) {
		value = d.value;
	}

	private ProtocolDouble(int v) {
		value = v;
	}

	public double asDouble() {
		return this.value / MULTIPLE;
	}

	public int GetValue() {
		return value;
	}

	public String Serialize() {
		return String.valueOf(value);
	}

	@Override
	public String toString() {
		return "PD"+String.valueOf(value);
	}
}
